### Getting started with pandas

> To load the pandas package and start working with it, import the package. The community agreed alias for pandas is `pd`, so loading pandas as `pd` is assumed standard practice for all of the pandas documentation.

Pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real-world data analysis in Python. Additionally, it has the broader goal of becoming the most powerful and flexible open source data analysis/manipulation tool available in any language. It is already well on its way toward this goal.

pandas is well suited for many different kinds of data:

* Tabular data with heterogeneously-typed columns, as in an SQL table or Excel spreadsheet
* Ordered and unordered (not necessarily fixed-frequency) time series data.
* Arbitrary matrix data (homogeneously typed or heterogeneous) with row and column labels
* Any other form of observational / statistical data sets. The data need not be labeled at all to be placed into a pandas data structure



In [1]:
import pandas as pd

In [3]:
from urllib.request import urlretrieve

In [5]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url, './test/italy-covid-daywise.csv')

('./test/italy-covid-daywise.csv', <http.client.HTTPMessage at 0x1660e02ed00>)

In [6]:
covid_df = pd.read_csv('./test/italy-covid-daywise.csv')

####   Data frames:
Data from the file is read and stored in a `DataFrame` object - one of the core data structures in Pandas for storing and working with tabular data. We typically use the `_df` suffix in the variable names for dataframes.

![dataframe]( https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRbnRxA5uwpq1aGMdzijo_Ribv5ymoo_u13aVCa3T3Ak9G3n_LutNZQKrCqR2fi6arUEnM&usqp=CAU)

In [8]:
type(covid_df)

pandas.core.frame.DataFrame

In [7]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

Looking at the dataframe, we can analyse the following things:
* The file contains the details of the pandemic in Italy
* The table maintains following records: new cases, new deaths and new tests
* The records are maintained for a total of 248 days, `starting 31-12-2019`

> **Various functions that can be performed on python data frames:**  
* `pd.read_csv` - Read data from a CSV file into a Pandas `DataFrame` object
* `.info()` - View basic infomation about rows, columns & data types
* `.describe()` - View statistical information about numeric columns
* `.columns` - Get the list of column names
* `.shape` - Get the number of rows & columns as a tuple

In [9]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [10]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

In [13]:
covid_df.shape

(248, 4)

In [14]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

### Retrieving data from a data frame

 Conceptually, you can think of a dataframe as a dictionary of lists: keys are column names, and values are lists/arrays containing data for the respective columns.

```Python
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}
```

Representing data in the above format has a few benefits:

* All values in a column typically have the same type of value, so it's more efficient to store them in a single array.
* Retrieving the values for a particular row simply requires extracting the elements at a given index from each column array.
* The representation is more compact (column names are recorded only once) compared to other formats that use a dictionary for each row of data

In [15]:
covid_df["new_cases"]

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

Each column is represented using a data structure called `Series`.
* Just like arrays, we can retrieve a specific value from the series (we can use `.at` as well
* Instead of using the indexing notation `[]`, Pandas also allows accessing columns as properties of the dataframe using the `.` notation. 
* We can also pass a list of coluumns within the indexing notation `[]` to retrieve a subset of the dataset.
* To access a specific row of data, Pandas provides the `.loc` method.

In [23]:
type(covid_df["new_cases"])

pandas.core.series.Series

In [24]:
covid_df["new_cases"][237]

1209.0

In [25]:
covid_df.at[237, "new_cases"]

1209.0

In [27]:
covid_df.new_cases.at[237]

1209.0

In [37]:
cases = covid_df[["new_cases", "date"]]

In [38]:
cases

new_cases        date
0          0.0  2019-12-31
1          0.0  2020-01-01
2          0.0  2020-01-02
3          0.0  2020-01-03
4          0.0  2020-01-04
..         ...         ...
243     1444.0  2020-08-30
244     1365.0  2020-08-31
245      996.0  2020-09-01
246      975.0  2020-09-02
247     1326.0  2020-09-03

[248 rows x 2 columns]

<b>Note:</b> 

The new data frame cases_df is simply a "view" of the original data frame covid_df. Both point to the same data in the computer's memory.

P.S: Sometimes you might need a full copy of the data frame, in which case you can use the copy method. 

In [39]:
covid_df.loc[230]

date          2020-08-17
new_cases          477.0
new_deaths           4.0
new_tests        21379.0
Name: 230, dtype: object

In [40]:
type(covid_df.loc[243])

pandas.core.series.Series

In [47]:
covid_df.head(3)

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN

In [46]:
covid_df.tail(3)

date  new_cases  new_deaths  new_tests
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

**We can find the first index that doesn't contain a NaN value using a column's first_valid_index method.**


In [48]:
covid_df.new_tests.first_valid_index()

111

In [51]:
covid_df.sample(5)

date  new_cases  new_deaths  new_tests
60   2020-02-29      238.0         4.0        NaN
155  2020-06-03      318.0        55.0    20035.0
52   2020-02-21        0.0         0.0        NaN
20   2020-01-20        0.0         0.0        NaN
21   2020-01-21        0.0         0.0        NaN

#### Various new functions used
- `covid_df['new_cases']` - Retrieving columns as a `Series` using the column name
- `new_cases[243]` - Retrieving values from a `Series` using an index
- `covid_df.at[243, 'new_cases']` - Retrieving a single value from a data frame
- `covid_df.copy()` - Creating a deep copy of a data frame
- `covid_df.loc[243]` - Retrieving a row or range of rows of data from the data frame
- `head`, `tail`, and `sample` - Retrieving multiple rows of data from the data frame
- `covid_df.new_tests.first_valid_index` - Finding the first non-empty index in a series


## Analysing the dataset further..

> **After covering various** `pandas` functions that can be performed on a dataset, we try to analyse and try to answer various questions usingh the data from the dataset.

* **Q1:** What are the total number of reported cases and deaths related to Covid-19 in Italy?
* **Q2:** What is the overall death rate (ratio of reported deaths to reported cases)?
* **Q3:** What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were reported.
* **Q4:** What fraction of tests returned a positive result?

In [52]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [56]:
#Q1
reported_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()
print('The Total number of reported cases were {} and the number of deaths reported were {}.'.format(int(reported_cases), int(total_deaths)) )

The Total number of reported cases were 271515 and the number of deaths reported were 35497.


In [67]:
#Q2
print('The overall Death rate reported: {:.2f}'.format(reported_cases/total_deaths) )

The overall Death rate reported: 7.65


In [74]:
#Q3
initial_tests = 935310
print("The total number of tests that were conducted till 2020-09-03:\n", int(initial_tests + covid_df.new_tests.sum()))


The total number of tests that were conducted till 2020-09-03:
 5214766


In [94]:
#Q4
positive_rate = reported_cases/ (covid_df.new_tests.sum()+initial_tests)
type(positive_rate)

numpy.float64

In [96]:
print('{:.2f}% were diagnosed positive'.format(positive_rate*100))

5.21% were diagnosed positive


## Querying and sorting rows in a dataset
Suppose we only want to look at the days which had more than 900 reported cases. We can use a boolean expression to check which rows satisfy this criterion.

In [101]:
high_cases = covid_df.new_cases >900

The boolean expression above returns a series containing `True` and `False` boolean values:

In [102]:
high_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245     True
246     True
247     True
Name: new_cases, Length: 248, dtype: bool

> Selecting a subset of rows from the original dataframe, corresponding to the True values in the series:

In [103]:
covid_df[high_cases]

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
71   2020-03-11      977.0       167.0        NaN
72   2020-03-12     2313.0       196.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[78 rows x 4 columns]

Using the `display()` function from `IPython.display' to change the default display settings.     

In [104]:
from IPython.display import display
with pd.option_context('display.max_rows', 100):
    display(covid_df[covid_df.new_cases > 1000])

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
74   2020-03-14     2547.0       252.0        NaN
75   2020-03-15     3497.0       173.0        NaN
76   2020-03-16     2823.0       370.0        NaN
77   2020-03-17     4000.0       347.0        NaN
78   2020-03-18     3526.0       347.0        NaN
79   2020-03-19     4207.0       473.0        NaN
80   2020-03-20     5322.0       429.0        NaN
81   2020-03-21     5986.0       625.0        NaN
82   2020-03-22     6557.0       795.0        NaN
83   2020-03-23     5560.0       649.0        NaN
84   2020-03-24     4789.0       601.0        NaN
85   2020-03-25     5249.0       743.0        NaN
86   2020-03-26     5210.0       685.0        NaN
87   2020-03-27     6153.0       660.0        NaN
88   2020-03-28     5959.0       971.0        NaN
89   2020-03-29     5974.0       887.0        NaN
90   2020-03-30     5217.0       758.0        NaN
91   2020-03-31     4050.0       810.0        NaN
92   2020-04-01     4053.0       839.0        NaN
93   2020-04-02     4782.0       727.0        NaN
94   2020-04-03     4668.0       760.0        NaN
95   2020-04-04     4585.0       764.0        NaN
96   2020-04-05     4805.0       681.0        NaN
97   2020-04-06     4316.0       527.0        NaN
98   2020-04-07     3599.0       636.0        NaN
99   2020-04-08     3039.0       604.0        NaN
100  2020-04-09     3836.0       540.0        NaN
101  2020-04-10     4204.0       612.0        NaN
102  2020-04-11     3951.0       570.0        NaN
103  2020-04-12     4694.0       619.0        NaN
104  2020-04-13     4092.0       431.0        NaN
105  2020-04-14     3153.0       564.0        NaN
106  2020-04-15     2972.0       604.0        NaN
107  2020-04-16     2667.0       578.0        NaN
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
115  2020-04-24     2646.0       464.0    95273.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
119  2020-04-28     1739.0       333.0    37554.0
120  2020-04-29     2091.0       382.0    38589.0
121  2020-04-30     2086.0       323.0    41441.0
122  2020-05-01     1872.0       285.0    43732.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
126  2020-05-05     1221.0       195.0    32211.0
127  2020-05-06     1075.0       236.0    37771.0
128  2020-05-07     1444.0       369.0    13665.0
129  2020-05-08     1401.0       274.0    45428.0
130  2020-05-09     1327.0       243.0    36091.0
131  2020-05-10     1083.0       194.0    31384.0
134  2020-05-13     1402.0       172.0    37049.0
236  2020-08-23     1071.0         3.0    47463.0
237  2020-08-24     1209.0         7.0    33358.0
240  2020-08-27     1366.0        13.0    57640.0
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

let's try to determine the days when the ratio of cases reported to tests conducted is higher than the overall `positive_rate.`

In [105]:
high_ratio_df = covid_df[covid_df.new_cases / covid_df.new_tests > positive_rate]

In [106]:
high_ratio_df

date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
120  2020-04-29     2091.0       382.0    38589.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
128  2020-05-07     1444.0       369.0    13665.0

In [108]:
#The result of performing an operation on two columns is a new series.
covid_df.new_cases / covid_df.new_tests


0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
243    0.026970
244    0.032055
245    0.018311
246         NaN
247         NaN
Length: 248, dtype: float64

We can use this series to add a new column to the data frame.

In [115]:
covid_df['positive_rate'] = (covid_df.new_cases / covid_df.new_tests)*100

In [116]:
covid_df

date  new_cases  new_deaths  new_tests  positive_rate
0    2019-12-31        0.0         0.0        NaN            NaN
1    2020-01-01        0.0         0.0        NaN            NaN
2    2020-01-02        0.0         0.0        NaN            NaN
3    2020-01-03        0.0         0.0        NaN            NaN
4    2020-01-04        0.0         0.0        NaN            NaN
..          ...        ...         ...        ...            ...
243  2020-08-30     1444.0         1.0    53541.0       2.696999
244  2020-08-31     1365.0         4.0    42583.0       3.205505
245  2020-09-01      996.0         6.0    54395.0       1.831051
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

[248 rows x 5 columns]

let's try removing the `positive_rate` column using the `drop` method.
To check all the parameters that can be passed onto the `drop` method, run help on it! `covid_df.drop?`

In [119]:
covid_df.drop(columns = 'positive_rate',inplace= True)


### Sorting rowsusing column values in a dataset

We can use `.sort_values` to sort the values of a row using a specific column. 
> In the below case, we sort the rows in a descending order, trying to fetch the highest number of cases in a single day. `.head` to fetch the top 10 days with highest number of recorded cases. The `ascending` parameter is set to False to sort the values in descending order

In [125]:
covid_df.sort_values('new_cases', ascending= False).head(10)

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN
83  2020-03-23     5560.0       649.0        NaN
80  2020-03-20     5322.0       429.0        NaN
85  2020-03-25     5249.0       743.0        NaN
90  2020-03-30     5217.0       758.0        NaN
86  2020-03-26     5210.0       685.0        NaN

Below, we try to sort the rows in accordance to the highest number of deaths recorded. 

In [129]:
covid_df.sort_values('new_deaths', ascending= False).head(10)

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN
95  2020-04-04     4585.0       764.0        NaN
94  2020-04-03     4668.0       760.0        NaN
90  2020-03-30     5217.0       758.0        NaN
85  2020-03-25     5249.0       743.0        NaN
93  2020-04-02     4782.0       727.0        NaN

It seems like the number of deaths increased just a week after the surge in the number of daily cases.

In [131]:
covid_df.sort_values('new_cases')

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
0    2019-12-31        0.0         0.0        NaN
29   2020-01-29        0.0         0.0        NaN
30   2020-01-30        0.0         0.0        NaN
32   2020-02-01        0.0         0.0        NaN
..          ...        ...         ...        ...
88   2020-03-28     5959.0       971.0        NaN
89   2020-03-29     5974.0       887.0        NaN
81   2020-03-21     5986.0       625.0        NaN
87   2020-03-27     6153.0       660.0        NaN
82   2020-03-22     6557.0       795.0        NaN

[248 rows x 4 columns]

Here, we tried looking at the days where the number of new cases registered was the least. It seems like we have a negative entry on `2020-06-20`.
Let's assume this was indeed a data entry error. We can use one of the following approaches for dealing with the missing or faulty value:
* Replace it with 0.
* Replace it with the average of the entire column
* Replace it with the average of the values on the previous & next date
* Discard the row entirely

 In this case, since we are dealing with data ordered by date, we can go ahead with the third approach.

In [134]:
covid_df.at[172, 'new_cases'] = covid_df.at[171, 'new_cases'] + covid_df.at[173, 'new_cases'] / 2

In [135]:
covid_df.loc[169:175]

date  new_cases  new_deaths  new_tests
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20      463.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

Entry of the new Functions used:
- `covid_df.new_cases.sum()` - Computing the sum of values in a column or series
- `covid_df[covid_df.new_cases > 1000]` - Querying a subset of rows satisfying the chosen criteria using boolean expressions
- `df['pos_rate'] = df.new_cases/df.new_tests` - Adding new columns by combining data from existing columns
- `covid_df.drop('positive_rate')` - Removing one or more columns from the data frame
- `sort_values` - Sorting the rows of a data frame using column values
- `covid_df.at[172, 'new_cases'] = ...` - Replacing a value within the data frame

### Working with dates in pandas
The data type of date in data frame `covid_df` is currently `object`, so Pandas does not know that this column is a date. We convert it into a `datetime` column using the `pd.to_datetime` method.

In [136]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

In [137]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [138]:
covid_df.date

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

 We can now extract different parts of the data into separate columns, using the `DatetimeIndex` class

In [140]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [141]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5
..         ...        ...         ...        ...   ...    ...  ...      ...
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3

[248 rows x 8 columns]

We check the overall metrics for June. We query the rows for Rows, choose a subset of columns, and use the `sum` method to aggregate each selected column's values.

In [146]:
covid_df_june = covid_df[covid_df.month == 6]

In [147]:
covid_df_june

date  new_cases  new_deaths  new_tests  year  month  day  weekday
153 2020-06-01      355.0        75.0    18053.0  2020      6    1        0
154 2020-06-02      178.0        60.0    25628.0  2020      6    2        1
155 2020-06-03      318.0        55.0    20035.0  2020      6    3        2
156 2020-06-04      321.0        71.0    27451.0  2020      6    4        3
157 2020-06-05      177.0        88.0    40470.0  2020      6    5        4
158 2020-06-06      518.0        85.0    34036.0  2020      6    6        5
159 2020-06-07      270.0        72.0    27894.0  2020      6    7        6
160 2020-06-08      197.0        53.0    16301.0  2020      6    8        0
161 2020-06-09      280.0        65.0    32200.0  2020      6    9        1
162 2020-06-10      283.0        79.0    37865.0  2020      6   10        2
163 2020-06-11      202.0        71.0    32991.0  2020      6   11        3
164 2020-06-12      379.0        53.0    37651.0  2020      6   12        4
165 2020-06-13      163.0        56.0    32880.0  2020      6   13        5
166 2020-06-14      346.0        78.0    29545.0  2020      6   14        6
167 2020-06-15      338.0        44.0    17463.0  2020      6   15        0
168 2020-06-16      301.0        26.0    27762.0  2020      6   16        1
169 2020-06-17      210.0        34.0    33957.0  2020      6   17        2
170 2020-06-18      328.0        43.0    32921.0  2020      6   18        3
171 2020-06-19      331.0        66.0    28570.0  2020      6   19        4
172 2020-06-20      463.0        47.0    29875.0  2020      6   20        5
173 2020-06-21      264.0        49.0    24581.0  2020      6   21        6
174 2020-06-22      224.0        24.0    16152.0  2020      6   22        0
175 2020-06-23      221.0        23.0    23225.0  2020      6   23        1
176 2020-06-24      113.0        18.0    30237.0  2020      6   24        2
177 2020-06-25      577.0       -31.0    29421.0  2020      6   25        3
178 2020-06-26      296.0        34.0    28331.0  2020      6   26        4
179 2020-06-27      255.0        30.0    29721.0  2020      6   27        5
180 2020-06-28      175.0         8.0    21183.0  2020      6   28        6
181 2020-06-29      174.0        22.0    15484.0  2020      6   29        0
182 2020-06-30      126.0         6.0    28471.0  2020      6   30        1

In [150]:
covid_df_june_subset = covid_df_june[['new_deaths', 'new_cases', 'new_tests']]

In [152]:
covid_df_june_subset.sum()

new_deaths      1404.0
new_cases       8383.0
new_tests     830354.0
dtype: float64

Let's try analyzing  if the number of cases reported on Sundays is higher than the average number of cases reported on Saturdays.

In [155]:
#Average number of cases on Saturdays
covid_df[covid_df.weekday == 5].new_cases.mean()

1203.2571428571428

In [156]:
#Average number of cases on Sundays
covid_df[covid_df.weekday == 6].new_cases.mean()

1247.2571428571428

From the above analysis, the average reported cases on sundays are more when compared to the average of cases reported on Saturdays.

### Grouping and aggergation of data
> Now, let's try summarizing the datwie data and crate a new dataframe with monthwise data using the `groupby` function.
* We group the data month wise, select specific columns and aggregate them using the `sum`.

In [167]:
covid_df_aggre = covid_df.groupby('month')[['new_cases', 'new_tests',  'new_deaths']] 

In [168]:
covid_df_aggre

In [169]:
covid_df_aggre.sum()

new_cases  new_tests  new_deaths
month                                  
1            3.0        0.0         0.0
2          885.0        0.0        21.0
3       100851.0        0.0     11570.0
4       101852.0   419591.0     16091.0
5        29073.0  1078720.0      5658.0
6         8383.0   830354.0      1404.0
7         6722.0   797692.0       388.0
8        21060.0  1098704.0       345.0
9         3297.0    54395.0        20.0
12           0.0        0.0         0.0

A new data frame that uses unique values from the column passed to `groupby` as the index is formed. Grouping and aggregation is a powerful method for progressively summarizing data into smaller data frames.

Instead of aggregating by sum, we also aggregate `mean`. Let's compute the average number of daily new cases, deaths, and tests for each month.

In [170]:
daily_df = covid_df.groupby('weekday')[['new_cases', 'new_tests',  'new_deaths']].mean()

In [171]:
daily_df

new_cases     new_tests  new_deaths
weekday                                       
0        1109.142857  19826.700000  124.800000
1         918.638889  31166.900000  129.944444
2         927.305556  34665.210526  144.750000
3        1095.666667  34488.578947  141.472222
4        1189.714286  39225.684211  144.885714
5        1203.257143  34589.473684  157.000000
6        1247.257143  28588.105263  159.457143

Apart from grouping, another form of aggregation is the running or cumulative sum of cases, tests, or death up to each row's date. We use `cumsum()` method to find out the cummilative sum till a row.

In [172]:
covid_df['total_cases'] = covid_df.new_cases.cumsum()
covid_df['total_deaths)'] = covid_df.new_deaths.cumsum()
covid_df['total_tests'] = covid_df.new_tests.cumsum() +initial_tests

In [174]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths)  total_tests  
0            0.0            0.0          NaN  
1            0.0            0.0          NaN  
2            0.0            0.0          NaN  
3            0.0            0.0          NaN  
4            0.0            0.0          NaN  
..           ...            ...          ...  
243     267464.0        35473.0    5117788.0  
244     268829.0        35477.0    5160371.0  
245     269825.0        35483.0    5214766.0  
246     270800.0        35491.0          NaN  
247     272126.0        35497.0          NaN  

[248 rows x 11 columns]

### Merging data from multiple sources
We can also merge two different data frames to determine other matricies. Here, we try downloading a new csv `locations.csv` that contains health relatedinformation for countries across the globe.

In [176]:
urlretrieve('https://gist.githubusercontent.com/aakashns/8684589ef4f266116cdce023377fc9c8/raw/99ce3826b2a9d1e6d0bde7e9e559fc8b6e9ac88b/locations.csv', 
            './test/locations.csv')

('./test/locations.csv', <http.client.HTTPMessage at 0x1661360e4c0>)

In [177]:
countries_df = pd.read_csv('./test/locations.csv')

In [178]:
countries_df

location continent    population  life_expectancy  \
0      Afghanistan      Asia  3.892834e+07            64.83   
1          Albania    Europe  2.877800e+06            78.57   
2          Algeria    Africa  4.385104e+07            76.88   
3          Andorra    Europe  7.726500e+04            83.73   
4           Angola    Africa  3.286627e+07            61.15   
..             ...       ...           ...              ...   
207          Yemen      Asia  2.982597e+07            66.12   
208         Zambia    Africa  1.838396e+07            63.89   
209       Zimbabwe    Africa  1.486293e+07            61.49   
210          World       NaN  7.794799e+09            72.58   
211  International       NaN           NaN              NaN   

     hospital_beds_per_thousand  gdp_per_capita  
0                         0.500        1803.987  
1                         2.890       11803.431  
2                         1.900       13913.839  
3                           NaN             NaN  
4                           NaN        5819.495  
..                          ...             ...  
207                       0.700        1479.147  
208                       2.000        3689.251  
209                       1.700        1899.775  
210                       2.705       15469.207  
211                         NaN             NaN  

[212 rows x 6 columns]

The downloaded CSV has data from 212 counties. Let's try looking for Italy in the dataset.

In [182]:
countries_df[countries_df.location == 'Italy']

location continent  population  life_expectancy  \
97    Italy    Europe  60461828.0            83.51   

    hospital_beds_per_thousand  gdp_per_capita  
97                        3.18       35220.084

Now, in order to mege the two data frames, we need at least one common column. Let's insert a `location` column in the `covid_df` dataframe with all values set to `"Italy"`.

In [184]:
covid_df['location'] = 'Italy'

In [185]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths)  total_tests location  
0            0.0            0.0          NaN    Italy  
1            0.0            0.0          NaN    Italy  
2            0.0            0.0          NaN    Italy  
3            0.0            0.0          NaN    Italy  
4            0.0            0.0          NaN    Italy  
..           ...            ...          ...      ...  
243     267464.0        35473.0    5117788.0    Italy  
244     268829.0        35477.0    5160371.0    Italy  
245     269825.0        35483.0    5214766.0    Italy  
246     270800.0        35491.0          NaN    Italy  
247     272126.0        35497.0          NaN    Italy  

[248 rows x 12 columns]

Now, we can try merging the two dataframes using the `.merge` method

In [188]:
complete_set = covid_df.merge(countries_df, on ='location')

In [189]:
complete_set

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths)  total_tests location continent  population  \
0            0.0            0.0          NaN    Italy    Europe  60461828.0   
1            0.0            0.0          NaN    Italy    Europe  60461828.0   
2            0.0            0.0          NaN    Italy    Europe  60461828.0   
3            0.0            0.0          NaN    Italy    Europe  60461828.0   
4            0.0            0.0          NaN    Italy    Europe  60461828.0   
..           ...            ...          ...      ...       ...         ...   
243     267464.0        35473.0    5117788.0    Italy    Europe  60461828.0   
244     268829.0        35477.0    5160371.0    Italy    Europe  60461828.0   
245     269825.0        35483.0    5214766.0    Italy    Europe  60461828.0   
246     270800.0        35491.0          NaN    Italy    Europe  60461828.0   
247     272126.0        35497.0          NaN    Italy    Europe  60461828.0   

     life_expectancy  hospital_beds_per_thousand  gdp_per_capita  
0              83.51                        3.18       35220.084  
1              83.51                        3.18       35220.084  
2              83.51                        3.18       35220.084  
3              83.51                        3.18       35220.084  
4              83.51                        3.18       35220.084  
..               ...                         ...             ...  
243            83.51                        3.18       35220.084  
244            83.51                        3.18       35220.084  
245            83.51                        3.18       35220.084  
246            83.51                        3.18       35220.084  
247            83.51                        3.18       35220.084  

[248 rows x 17 columns]

### Writing back to files
So, before tring to save the data to a file, we may want to remove the unnecessary columns from the data frame.
* We can either use the `.drop method`
* Or, we can load the required columns to a new data frame and save that data frame to the file.

In [194]:
complete_set.drop(columns = ['month', 'day', 'weekday'])

date  new_cases  new_deaths  new_tests  year  total_cases  \
0   2019-12-31        0.0         0.0        NaN  2019          0.0   
1   2020-01-01        0.0         0.0        NaN  2020          0.0   
2   2020-01-02        0.0         0.0        NaN  2020          0.0   
3   2020-01-03        0.0         0.0        NaN  2020          0.0   
4   2020-01-04        0.0         0.0        NaN  2020          0.0   
..         ...        ...         ...        ...   ...          ...   
243 2020-08-30     1444.0         1.0    53541.0  2020     267464.0   
244 2020-08-31     1365.0         4.0    42583.0  2020     268829.0   
245 2020-09-01      996.0         6.0    54395.0  2020     269825.0   
246 2020-09-02      975.0         8.0        NaN  2020     270800.0   
247 2020-09-03     1326.0         6.0        NaN  2020     272126.0   

     total_deaths)  total_tests location continent  population  \
0              0.0          NaN    Italy    Europe  60461828.0   
1              0.0          NaN    Italy    Europe  60461828.0   
2              0.0          NaN    Italy    Europe  60461828.0   
3              0.0          NaN    Italy    Europe  60461828.0   
4              0.0          NaN    Italy    Europe  60461828.0   
..             ...          ...      ...       ...         ...   
243        35473.0    5117788.0    Italy    Europe  60461828.0   
244        35477.0    5160371.0    Italy    Europe  60461828.0   
245        35483.0    5214766.0    Italy    Europe  60461828.0   
246        35491.0          NaN    Italy    Europe  60461828.0   
247        35497.0          NaN    Italy    Europe  60461828.0   

     life_expectancy  hospital_beds_per_thousand  gdp_per_capita  
0              83.51                        3.18       35220.084  
1              83.51                        3.18       35220.084  
2              83.51                        3.18       35220.084  
3              83.51                        3.18       35220.084  
4              83.51                        3.18       35220.084  
..               ...                         ...             ...  
243            83.51                        3.18       35220.084  
244            83.51                        3.18       35220.084  
245            83.51                        3.18       35220.084  
246            83.51                        3.18       35220.084  
247            83.51                        3.18       35220.084  

[248 rows x 14 columns]

We use `to_csv` to write the data to the csv file. The `to_csv` function also includes an additional column for storing the index of the dataframe by default. We pass `index=None` to turn off this behavior.

In [200]:
complete_set.to_csv('./test/processed.csv', index = None)

We can try verifying whether the csv is created

In [201]:
import os

In [202]:
os.listdir('./test')

['emis2.txt',
 'italy-covid-daywise.csv',
 'loan1.txt',
 'loan2.txt',
 'loan3.txt',
 'locations.csv',
 'processed.csv']

We can see that `processed.csv` exists in the directory.